internal/services/device_status_ingest_service.go

```diff
@@ -261,25 +275,67 @@ func (i *DeviceStatusIngestService) processOdometer(datum *models.UserDeviceDatu
 	}
 
 	var oldOdometer null.Float64
+	var oldOdometerTimestamp null.Time
 	if datum.Signals.Valid {
 		if o, err := extractOdometer(datum.Signals.JSON); err == nil {
 			oldOdometer = null.Float64From(o)
 		}
+		if t, err := extractOdometerTime(datum.Signals.JSON); err == nil {
+			oldOdometerTimestamp = null.TimeFrom(t)
+		}
 	}
 
 	now := time.Now()
-	odometerOffCooldown := !datum.LastOdometerEventAt.Valid || now.Sub(datum.LastOdometerEventAt.Time) >= odometerCooldown
+	odometerOffCooldown := !oldOdometerTimestamp.Valid || now.Sub(oldOdometerTimestamp.Time) >= odometerCooldown
 	odometerChanged := !oldOdometer.Valid || newOdometer.Float64 > oldOdometer.Float64
 
 	if odometerOffCooldown && odometerChanged {
-		datum.LastOdometerEventAt = null.TimeFrom(now)
-		if newOdometer.Float64 > 0.01 {
-			// Since this function will always receive 0.0 for odo if not present
-			// if odometer value is 0 then it must have been fake
-			datum.RealLastOdometerEventAt = null.TimeFrom(now)
-		}
+		oldOdometerTimestamp = null.TimeFrom(now)
 		i.emitOdometerEvent(device, dd, integrationID, newOdometer.Float64)
 	}
+}
+
+func (i *DeviceStatusIngestService) processLastLocation(datum *models.UserDeviceDatum, newLastLocation null.Float64) {
+	if !newLastLocation.Valid {
+		return
+	}
+	var oldLastLocation null.Float64
+
+	if datum.Signals.Valid {
+		if o, err := extractLastLocation(datum.Signals.JSON); err == nil {
+			oldLastLocation = null.Float64From(o)
+		}
+	}
+
+	now := time.Now()
+
+	locationChanged := !oldLastLocation.Valid || newLastLocation.Float64 > oldLastLocation.Float64
+
+	if locationChanged {
+		datum.LastLocationEventAt = null.TimeFrom(now)
+	}
+}
+
+func (i *DeviceStatusIngestService) processObd2(datum *models.UserDeviceDatum, newObd2 null.Float64) {
+	if !newObd2.Valid {
+		return
+	}
+
+	var oldObd2 null.Float64
+
+	if datum.Signals.Valid {
+		if o, err := extractLastLocation(datum.Signals.JSON); err == nil {
```

Comment:
this should be a new method `extractObd2Signal` or just something like `obd2SignalExists`.

at the end you're just checking if a few different obd2 signals exist in the json